### PR TITLE
Load autocompletions from directory

### PR DESCRIPTION
**Pull request by funkyfuture**
this adds the capability to define directories to load autocompletions from.

i started to move platform-specifics to a new module. this isn't finished yet, though working. i will finish it if there is approving feedback.

unfortunatley i can't rebase onto master as git won't stop to autoconvert `scripts/xonsh.bat` what i also try to tackle in `.gitattributes`. without success.

### CODE REVIEW

**Comment by gforsyth** [comment]
Hey @funkyfuture, thanks for putting this in.  

I haven't had the issues you describe with `xonsh.bat` but hopefully someone else can pipe up with a solution.  

Could you split this up into several PRs?  I see the autocompletion loading, the creation of `xonsh.platform` (which you mention) and the removal of some old Py2 compatibility stuff.  I think those are all good ideas but they should be treated separately.

**Comment by GuillaumeLeclerc** [comment]
I'm having the same problem with `xonsh.bat`, but I only do merge no rebase and it seems fine

**Comment by funkyfuture** [comment]
@gforsyth yeah, it doesn't seem to introduce conflicts. will do this europe's evening.

**Comment by funkyfuture** [comment]
updated the commits.

@GuillaumeLeclerc i have git 2.8.2 on Manjalo Linux

**Comment by funkyfuture** [comment]
sorry i can't squash in the missing `from pathlib import Path`. would be great if someone did this locally.

**Comment by gforsyth** [comment]
@funkyfuture -- it's fine to just add another commit to fix the missing import.  Rebasing is great but don't let the great be the enemy of the good.

**Comment by scopatz** [comment]
Yeah @funkyfuture, to second @gforsyth, feel free to merge.

**Comment by melund** [comment]
There are some tests failing now

**Comment by funkyfuture** [comment]
fixed it.

File: xonsh/completer.py [hunk 1]
```diff
@@ -425,38 +427,40 @@ def bash_complete(self, prefix, line, begidx, endidx):
...
+                for _file in (x for x in path.glob('*') if x.is_file()):
+                    sources.append('source ' + str(_file))
+        return sources
 
     def _load_bash_complete_funcs(self):
         self.bash_complete_funcs = bcf = {}
-        inp = self._source_completions()
-        if len(inp) == 0:
+        inp = self._collect_completions_sources()
+        if not inp:
```
Comment (scopatz) [inline xonsh/completer.py]: Please revert this to `if len(inp) == 0`. This is the more precise test, and ambiguity isn't needed here. Thanks!

File: xonsh/completer.py [hunk 2]
```diff
@@ -467,17 +471,16 @@ def _load_bash_complete_funcs(self):
             bcf[cmd] = m.group(1)
 
     def _load_bash_complete_files(self):
-        inp = self._source_completions()
-        if len(inp) == 0:
+        inp = self._collect_completions_sources()
+        if not inp:
```
Comment (scopatz) [inline xonsh/completer.py]: Same.

**Comment by gforsyth** [comment]
I can confirm that this is working on Arch.  There are weird completion behaviors with certain things (most notably `pacman`) but those were there already.  This looks good to merge to me once @scopatz changes have been made.

**Comment by scopatz** [comment]
Hey @funkyfuture - just a couple of small changes but otherwise lgtm

File: xonsh/completer.py (hunk 1 above)
Comment (funkyfuture) [inline xonsh/completer.py]: my rationale is that both expressions are equal as inp is always a list, so there is no ambiguity. my motivation is a better readability.

Comment (scopatz) [inline xonsh/completer.py]: I personally find the original way much more readable :) But I guess this way is much faster,

```console
scopatz@localhost ~ $ x = []
scopatz@localhost ~ $ timeit "len(x) > 0"
10000000 loops, best of 3: 75.8 ns per loop
scopatz@localhost ~ $ timeit "not x"
10000000 loops, best of 3: 28.4 ns per loop
```

Comment (funkyfuture) [inline xonsh/completer.py]: thanks for the research. never thought about that aspect. what results shows `not len(x)`? ( i currently bugged my xonsh)

Comment (scopatz) [inline xonsh/completer.py]: I am getting, 

```console
scopatz@localhost ~ $ timeit "not len(x)"
10000000 loops, best of 3: 64 ns per loop
```

**Comment by scopatz** [comment]
Ok @funkyfuture - I think this looks good.  Any other changes you want to make, or is it good to merge?

**Comment by funkyfuture** [comment]
gtm

**Comment by scopatz** [comment]
Thanks a million @funkyfuture! This is awesome to have.

**Pull request merged**